# Business Understanding
## Overview of the business
A company that is seeking to expand into the aviation industry and wants buy aircrafts for its new division

## Business Problem

Being new in the aviation industry the company is unaware of the potential risks that are associated with aircrafts 

## Business Objectives
The company wants to understand the potential risks it may face in the aviation industry
The company wants to obtain a list of recommended aircrafts that are considered to have low risks in regards to its operational use

## Success Criteria
To obtain a list of aircrafts that are considered to have low risks

To outline the potential risks the company may experience in the aviation industry

To produce relevant visualisations that provide insights on aircraft and the aviation industry

## Project plan

To go through a dataset that contains accidents that have occurred in the aviation industry and to obtain relevant insights that can be able to solve the business problem






# Familization with the dataset

In [60]:
# Importing the pandas and matplotlib libraries
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [61]:
pd.options.display.max_columns = None

Loading the dataset to be used


In [62]:
aviation_data_df = pd.read_csv("data/AviationData.csv", encoding='latin-1', dtype='str')


In [63]:
usstate_codes_df = pd.read_csv("data/USState_Codes.csv")

Checking the head, tail and a sample of the dataset so to get an understanding of the dataset we are working with


In [64]:
aviation_data_df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  Registration.Number      Make     Model Amateur.Built Number.of.Engines  \
0              NC6404   Stinson     108-3            No                 1   
1              N5069P     Piper  PA24-180            No                 1   
2              N5142R    Cessna      172M            No                 1   
3              N1168J  Rockwell       112            No                 1   
4               N15NY    Cessna       501            No               NaN   

     Engine.Type FAR.Description Schedule Purpose.of.flight Air.carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

  Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                    2                      0                    0   
1                    4                      0                    0   
2                    3                    NaN                  NaN   
3                    2                      0                    0   
4                    1                      2                  NaN   

  Total.Uninjured Weather.Condition Broad.phase.of.flight   Report.Status  \
0               0               UNK                Cruise  Probable Cause   
1               0               UNK               Unknown  Probable Cause   
2             NaN               IMC                Cruise  Probable Cause   
3               0               IMC                Cruise  Probable Cause   
4               0               VMC              Approach  Probable Cause   

  Publication.Date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

In [65]:
aviation_data_df.tail()

Event.Id Investigation.Type Accident.Number  Event.Date  \
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
88884  Annapolis, MD  United States      NaN       NaN          NaN   
88885    Hampton, NH  United States      NaN       NaN          NaN   
88886     Payson, AZ  United States  341525N  1112021W          PAN   
88887     Morgan, UT  United States      NaN       NaN          NaN   
88888     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
88884          NaN           Minor             NaN               NaN   
88885          NaN             NaN             NaN               NaN   
88886       PAYSON       Non-Fatal     Substantial          Airplane   
88887          NaN             NaN             NaN               NaN   
88888          NaN           Minor             NaN               NaN   

      Registration.Number                        Make      Model  \
88884              N1867H                       PIPER  PA-28-151   
88885              N2895Z                    BELLANCA       7ECA   
88886              N749PJ  AMERICAN CHAMPION AIRCRAFT      8GCBC   
88887              N210CU                      CESSNA       210N   
88888              N9026P                       PIPER  PA-24-260   

      Amateur.Built Number.of.Engines Engine.Type FAR.Description Schedule  \
88884            No               NaN         NaN             091      NaN   
88885            No               NaN         NaN             NaN      NaN   
88886            No                 1         NaN             091      NaN   
88887            No               NaN         NaN             091      NaN   
88888            No               NaN         NaN             091      NaN   

      Purpose.of.flight         Air.carrier Total.Fatal.Injuries  \
88884          Personal                 NaN                    0   
88885               NaN                 NaN                    0   
88886          Personal                 NaN                    0   
88887          Personal  MC CESSNA 210N LLC                    0   
88888          Personal                 NaN                    0   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
88884                      1                    0               0   
88885                      0                    0               0   
88886                      0                    0               1   
88887                      0                    0               0   
88888                      1                    0               1   

      Weather.Condition Broad.phase.of.flight Report.Status Publication.Date  
88884               NaN                   NaN           NaN       29-12-2022  
88885               NaN                   NaN           NaN              NaN  
88886               VMC                   NaN           NaN       27-12-2022  
88887               NaN                   NaN           NaN              NaN  
88888               NaN                   NaN           NaN       30-12-2022

In [66]:
aviation_data_df.sample(10)

Event.Id Investigation.Type Accident.Number  Event.Date  \
54235  20030310X00305           Accident      NYC03FA061  2003-03-02   
11992  20001214X36805           Accident      LAX85FA286  1985-06-19   
87959  20220616105269           Accident      ERA22LA269  2022-06-11   
10194  20001214X41491           Accident      ANC85LA019  1984-11-01   
15062  20001213X33929           Accident      LAX86LA270  1986-06-22   
57330  20040831X01336           Accident      SEA04MA167  2004-08-21   
40851  20001208X07047           Accident      IAD97LA015  1996-11-04   
76661  20150526X41555           Accident      GAA15CA088  2015-05-23   
38322  20001207X04459           Accident      CHI95LA321  1995-09-13   
74518  20131122X05100           Incident      DCA14IA016  2013-11-22   

                  Location        Country   Latitude    Longitude  \
54235         MONTEREY, MA  United States  42.210556   -73.262222   
11992     SQUAW VALLEY, CA  United States        NaN          NaN   
87959           Butler, PA  United States    404636N     0079574W   
10194        ANCHORAGE, AK  United States        NaN          NaN   
15062          AGUANGA, CA  United States        NaN          NaN   
57330  Battle Mountain, NV  United States    40.4525  -117.491944   
40851       UNION VALE, NY  United States        NaN          NaN   
76661           Sussex, NJ  United States    041122N     0743722W   
38322      BOYNE FALLS, MI  United States        NaN          NaN   
74518          Wichita, KS  United States    374451N     0971316W   

      Airport.Code                  Airport.Name Injury.Severity  \
54235          NaN                           NaN        Fatal(4)   
11992          NaN                           NaN        Fatal(2)   
87959          BTP        PITTSBURGH/BUTLER RGNL       Non-Fatal   
10194          MRI                       MERRILL       Non-Fatal   
15062          NaN                           NaN       Non-Fatal   
57330          NaN                           NaN        Fatal(5)   
40851          44N                     SKY ACRES       Non-Fatal   
76661          FWN                        SUSSEX       Non-Fatal   
38322          BFA                BOYNE MOUNTAIN       Non-Fatal   
74518          AAO  Colonel James Jabara Airport       Non-Fatal   

      Aircraft.damage Aircraft.Category Registration.Number  \
54235     Substantial               NaN              N4072R   
11992       Destroyed               NaN              N214RM   
87959     Substantial        Helicopter               N688G   
10194     Substantial               NaN              N5450E   
15062     Substantial               NaN              N3885A   
57330       Destroyed               NaN                N2YN   
40851     Substantial               NaN              N3903C   
76661     Substantial          Airplane              N9959V   
38322     Substantial               NaN              N76919   
74518             NaN          Airplane              N780BA   

                       Make               Model Amateur.Built  \
54235                 Piper           PA-32-300            No   
11992                  Bell              214B-1            No   
87959  WALTER S JOHNSON III                AR-1            No   
10194                 Helio               H-250            No   
15062             Schweizer           SGS 1-26C            No   
57330                  Bell                 407            No   
40851                 Ayres  SCORPION EXECUTIVE           Yes   
76661                CESSNA                172M            No   
38322                Cessna                 140            No   
74518                BOEING        747 - 409LCF            No   

      Number.of.Engines    Engine.Type FAR.Description Schedule  \
54235                 1  Reciprocating             NaN      NaN   
11992                 1    Turbo Shaft             NaN      UNK   
87959                 1  Reciprocating             091      NaN   
10194                 1  Reciproc

In [67]:
aviation_data_df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [68]:
aviation_data_df.shape

(88889, 31)

In [69]:
len(aviation_data_df.columns)

31

In [70]:
usstate_codes_df.head()

US_State Abbreviation
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA

In [71]:
usstate_codes_df.tail()

US_State Abbreviation
57  Virgin Islands           VI
58   Washington_DC           DC
59  Gulf of mexico           GM
60  Atlantic ocean           AO
61   Pacific ocean           PO

In [72]:
usstate_codes_df.shape

(62, 2)

In [73]:
usstate_codes_df.columns

Index(['US_State', 'Abbreviation'], dtype='object')

# DATA UNDERSTANDING



The data folder contains two data sets

## Aviation dataset
This dataset contains data of reported aviation accidents and information regarding the accident reported.

The dataset contains 31 columns with each column containing information on the accident reported.

Some of the columns in the dataset include:date of the accident, the severity of the accident, country where the accident occurred, aircraft damage, aircraft type and model etc

The dataset contains 88889 records

## US State Codes dataset
This dataset contains information on the full name of abbreviated us state codes

The dataset contains two columns i.e the US state and its abbrevaiation

The dataset contains 62 records





# Data cleaning

## Aviation Dataset

In [74]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Event.Id                88889 non-null  object
 1   Investigation.Type      88889 non-null  object
 2   Accident.Number         88889 non-null  object
 3   Event.Date              88889 non-null  object
 4   Location                88837 non-null  object
 5   Country                 88663 non-null  object
 6   Latitude                34382 non-null  object
 7   Longitude               34373 non-null  object
 8   Airport.Code            50249 non-null  object
 9   Airport.Name            52790 non-null  object
 10  Injury.Severity         87889 non-null  object
 11  Aircraft.damage         85695 non-null  object
 12  Aircraft.Category       32287 non-null  object
 13  Registration.Number     87572 non-null  object
 14  Make                    88826 non-null  object
 15  Mo

Using pd.info we can observe the following

a) every column contain string values

b)They are columns with null values

c)Columns names contain different structures and need to be standardized

### Converting column names to a simple standardized form

a)column names to be converted to lower case

b)column names to be seperated by underscore


In [75]:
#converting the columns into lower case
aviation_data_df.columns=aviation_data_df.columns.str.lower()

In [76]:
#seperating column names with an underscore rather than with a period

aviation_data_df.columns=aviation_data_df.columns.str.replace('.','_')

aviation_data_df.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

In [77]:
aviation_data_df.head()

event_id investigation_type accident_number  event_date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built number_of_engines  \
0              NC6404   Stinson     108-3            No                 1   
1              N5069P     Piper  PA24-180            No                 1   
2              N5142R    Cessna      172M            No                 1   
3              N1168J  Rockwell       112            No                 1   
4               N15NY    Cessna       501            No               NaN   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

  total_fatal_injuries total_serious_injuries total_minor_injuries  \
0                    2                      0                    0   
1                    4                      0                    0   
2                    3                    NaN                  NaN   
3                    2                      0                    0   
4                    1                      2                  NaN   

  total_uninjured weather_condition broad_phase_of_flight   report_status  \
0               0               UNK                Cruise  Probable Cause   
1               0               UNK               Unknown  Probable Cause   
2             NaN               IMC                Cruise  Probable Cause   
3               0               IMC                Cruise  Probable Cause   
4               0               VMC              Approach  Probable Cause   

  publication_date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

### Converting columns that contain string values to respective dtypes

This columns were all converted to string values while reading the csv so to clear error generated.Therefore we need to go through each column and with intuition convert some of the columns back to their original data type

In [78]:
aviation_data_df.sample(10)

event_id investigation_type accident_number  event_date  \
23935  20001213X29455           Accident      MKC89FA194  1989-09-02   
3737   20001214X42083           Accident      NYC83LA055  1983-01-21   
23181  20001213X28532           Accident      CHI89LA107  1989-06-11   
1516   20020917X02838           Accident      ATL82FA161  1982-06-11   
11314  20001214X36080           Incident      ATL85IA132  1985-04-12   
69604  20110306X84105           Accident      CEN11CA213  2011-03-01   
41751  20001208X07868           Accident      CHI97LA125  1997-05-10   
6582   20001214X44935           Accident      NYC84FFJ02  1983-10-16   
5137   20001214X43482           Accident      NYC83FA148  1983-06-22   
39218  20001208X05377           Accident      CHI96LA105  1996-03-12   

                   location        country latitude longitude airport_code  \
23935       LEES SUMMIT, MO  United States      NaN       NaN          K84   
3737            MERIDEN, CT  United States      NaN       NaN          MMK   
23181           NEW ERA, MI  United States      NaN       NaN          PVT   
1516             HAZARD, KY  United States      NaN       NaN          137   
11314            MARION, AL  United States      NaN       NaN          A09   
69604  Glenwood Springs, CO  United States  393022N  1071832W          GWS   
41751         DICKINSON, ND  United States      NaN       NaN          DIK   
6582         WILLIAMSON, NY  United States      NaN       NaN          NaN   
5137         GREAT BEND, PA  United States      NaN       NaN          NaN   
39218         CAMP LAKE, WI  United States      NaN       NaN          49C   

                     airport_name injury_severity aircraft_damage  \
23935                 LEES SUMMIT       Non-Fatal       Destroyed   
3737              MERIDEN/MARKHAM       Non-Fatal     Substantial   
23181                   LANKFER'S       Non-Fatal     Substantial   
1516                       HAZARD       Non-Fatal       Destroyed   
11314                PERRY COUNTY        Incident           Minor   
69604  Glenwood Springs Municipal       Non-Fatal     Substantial   
41751         DICKINSON MUNICIPAL       Non-Fatal     Substantial   
6582                          NaN       Non-Fatal     Substantial   
5137                     KIRKWOOD        Fatal(2)       Destroyed   
39218                   CAMP LAKE       Non-Fatal     Substantial   

      aircraft_category registration_number                         make  \
23935               NaN              N8705Z                       Cessna   
3737                NaN              N67897                       Cessna   
23181               NaN              N35561                       Cessna   
1516         Helicopter              N5781Z                 Aerospatiale   
11314               NaN               N46PC                        Piper   
69604        Helicopter              N74816  ROBINSON HELICOPTER COMPANY   
41751               NaN              N280TC                        Piper   
6582                NaN              N82526                        Piper   
5137                NaN              N2293Y                       Cessna   
39218               NaN              N2966U                       Cessna   

            model amateur_built number_of_engines    engine_type  \
23935        206A            No                 1  Reciprocating   
3737          152            No                 1  Reciprocating   
23181        172I            No                 1  Reciprocating   
1516       AS350D            No                 1    Turbo Shaft   
11314  PA-32R-300            No                 1  Reciprocating   
69604      R44 II            No                 1  Reciprocating   
41751   PA-28-140            No                 1  Reciprocating   
6582    PA-25-235            No                 1  Reciprocating   
5137          177            No                 1  Reciprocating   
39218         172            No                 1  Reciprocating   

               

#### number_of_engines,total_fatal_injuries,total_minor_injuries,total_serious_injuries and total_uninjured to interger values
Taking a sample of the dataset we can safely deduce that the following columns originally contained interger values

number_of_engines,total_fatal_injuries,total_minor_injuries,total_serious_injuries and total_uninjured

In [79]:
#We first convert the columns above to numeric
aviation_data_df['number_of_engines'] = pd.to_numeric(aviation_data_df['number_of_engines'],errors= 'coerce')
aviation_data_df['total_fatal_injuries'] = pd.to_numeric(aviation_data_df['total_fatal_injuries'], errors = 'coerce')
aviation_data_df['total_serious_injuries'] = pd.to_numeric(aviation_data_df['total_serious_injuries'], errors = 'coerce')
aviation_data_df['total_minor_injuries'] = pd.to_numeric(aviation_data_df['total_minor_injuries'], errors = 'coerce')
aviation_data_df['total_uninjured'] = pd.to_numeric(aviation_data_df['total_uninjured'], errors = 'coerce')


In [80]:

aviation_data_df[['number_of_engines','total_fatal_injuries','total_serious_injuries','total_minor_injuries','total_uninjured']]=aviation_data_df[['number_of_engines','total_fatal_injuries','total_serious_injuries','total_minor_injuries','total_uninjured']].astype('Int32')

In [81]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   event_id                88889 non-null  object
 1   investigation_type      88889 non-null  object
 2   accident_number         88889 non-null  object
 3   event_date              88889 non-null  object
 4   location                88837 non-null  object
 5   country                 88663 non-null  object
 6   latitude                34382 non-null  object
 7   longitude               34373 non-null  object
 8   airport_code            50249 non-null  object
 9   airport_name            52790 non-null  object
 10  injury_severity         87889 non-null  object
 11  aircraft_damage         85695 non-null  object
 12  aircraft_category       32287 non-null  object
 13  registration_number     87572 non-null  object
 14  make                    88826 non-null  object
 15  mo

In [82]:
aviation_data_df.head()

event_id investigation_type accident_number  event_date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

#### converting event_date and publication_date to pandas.Datetime

In [83]:
#Converting columns that contain dates into pandas.datetime
# that is event_date and publication_date


aviation_data_df['event_date']=pd.to_datetime(aviation_data_df['event_date'],format="%Y-%m-%d" ,yearfirst= True, errors= 'coerce')



In [84]:
aviation_data_df['publication_date']=pd.to_datetime(aviation_data_df['publication_date'],format="%d-%m-%Y", yearfirst= True, errors='coerce')

In [85]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [86]:
aviation_data_df[['event_date','publication_date']].sample(20)

event_date publication_date
41641 1997-04-24       1997-10-14
53038 2002-07-13       2002-12-06
10669 1985-01-10              NaT
27882 1991-05-03       1993-01-25
68290 2010-06-10       2020-09-25
12675 1985-08-16              NaT
53371 2002-08-26       2003-04-01
4080  1983-03-11              NaT
56666 2004-05-15       2006-01-31
33631 1993-08-25       1994-09-26
85519 2020-10-21       2021-08-20
55331 2003-08-19       2004-03-02
56351 2004-03-14       2004-12-28
69590 2011-02-26       2020-09-25
57087 2004-07-17       2005-07-07
69652 2011-03-12       2020-11-03
21517 1988-09-23       1989-09-25
69957 2011-05-13       2020-09-25
19242 1987-11-27       1988-12-27
29501 1991-11-25       1996-08-05

In [90]:
#It might be useful to extract the years and the months of the event_date and publication_date which can be useful when doing visualisations
aviation_data_df['event_year'] = aviation_data_df['event_date'].dt.year
aviation_data_df['event_month'] = aviation_data_df['event_date'].dt.month_name()
aviation_data_df['publication_year'] = aviation_data_df['publication_date'].dt.year
aviation_data_df['publication_month'] = aviation_data_df['publication_date'].dt.month_name()

In [93]:
aviation_data_df.head()

event_id investigation_type accident_number event_date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   
1  20001218X45447           Accident      LAX94LA336 1962-07-19   
2  20061025X01555           Accident      NYC07LA005 1974-08-30   
3  20001218X45448           Accident      LAX96LA321 1977-06-19   
4  20041105X01764           Accident      CHI79FA064 1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  event_year event_month  publication_year publication_month  
0              NaT        1948     October               NaN               NaN  
1       1996-09-19        1962        July            1996.0         September  
2       2007-02-26        1974      August            2007.0          February  
3       2000-09-12        1977        June            2000.0         September  
4       1980-04-16        1979      August            1980.0             April

In [92]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [94]:
#due to the null values present in the publication year we need to convert the publication year to interger

aviation_data_df['publication_year'] = aviation_data_df['publication_year'].astype('Int64')

In [98]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [102]:
#Let us reorder our new created columns to be closer to their referenced column
column_names=aviation_data_df.columns.to_list()
column_names

['event_id',
 'investigation_type',
 'accident_number',
 'event_date',
 'location',
 'country',
 'latitude',
 'longitude',
 'airport_code',
 'airport_name',
 'injury_severity',
 'aircraft_damage',
 'aircraft_category',
 'registration_number',
 'make',
 'model',
 'amateur_built',
 'number_of_engines',
 'engine_type',
 'far_description',
 'schedule',
 'purpose_of_flight',
 'air_carrier',
 'total_fatal_injuries',
 'total_serious_injuries',
 'total_minor_injuries',
 'total_uninjured',
 'weather_condition',
 'broad_phase_of_flight',
 'report_status',
 'publication_date',
 'event_year',
 'event_month',
 'publication_year',
 'publication_month']

In [104]:
reorderd_columns=['event_id',
 'investigation_type',
 'accident_number',
 'event_date',
 'event_year',
 'event_month',
 'location',
 'country',
 'latitude',
 'longitude',
 'airport_code',
 'airport_name',
 'injury_severity',
 'aircraft_damage',
 'aircraft_category',
 'registration_number',
 'make',
 'model',
 'amateur_built',
 'number_of_engines',
 'engine_type',
 'far_description',
 'schedule',
 'purpose_of_flight',
 'air_carrier',
 'total_fatal_injuries',
 'total_serious_injuries',
 'total_minor_injuries',
 'total_uninjured',
 'weather_condition',
 'broad_phase_of_flight',
 'report_status',
 'publication_date',
 'publication_year',
 'publication_month']

In [105]:
aviation_data_df=aviation_data_df[reorderd_columns]

In [106]:
aviation_data_df.head()

event_id investigation_type accident_number event_date  event_year  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24        1948   
1  20001218X45447           Accident      LAX94LA336 1962-07-19        1962   
2  20061025X01555           Accident      NYC07LA005 1974-08-30        1974   
3  20001218X45448           Accident      LAX96LA321 1977-06-19        1977   
4  20041105X01764           Accident      CHI79FA064 1979-08-02        1979   

  event_month         location        country   latitude   longitude  \
0     October  MOOSE CREEK, ID  United States        NaN         NaN   
1        July   BRIDGEPORT, CA  United States        NaN         NaN   
2      August    Saltville, VA  United States  36.922223  -81.878056   
3        June       EUREKA, CA  United States        NaN         NaN   
4      August       Canton, OH  United States        NaN         NaN   

  airport_code airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN          NaN        Fatal(2)       Destroyed               NaN   
1          NaN          NaN        Fatal(4)       Destroyed               NaN   
2          NaN          NaN        Fatal(3)       Destroyed               NaN   
3          NaN          NaN        Fatal(2)       Destroyed               NaN   
4          NaN          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  publication_year publication_month  
0              NaT              <NA>               NaN  
1       1996-09-19              1996         September  
2       2007-02-26              2007          February  
3       2000-09-12              2000         September  
4       1980-04-16              1980             April

#### Latitudes and longitude

In [115]:
#Let us check the percentage of null values in the latitude and longitude
aviation_data_df.isnull().mean()

event_id                  0.000000
investigation_type        0.000000
accident_number           0.000000
event_date                0.000000
event_year                0.000000
event_month               0.000000
location                  0.000585
country                   0.002542
latitude                  0.613203
longitude                 0.613304
airport_code              0.434699
airport_name              0.406113
injury_severity           0.011250
aircraft_damage           0.035932
aircraft_category         0.636772
registration_number       0.014816
make                      0.000709
model                     0.001035
amateur_built             0.001147
number_of_engines         0.068445
engine_type               0.079616
far_description           0.639742
schedule                  0.858453
purpose_of_flight         0.069660
air_carrier               0.812710
total_fatal_injuries      0.128261
total_serious_injuries    0.140737
total_minor_injuries      0.134246
total_uninjured     

In [119]:
aviation_data_df[['location','country','latitude','longitude']].sample(30)

location        country   latitude    longitude
58194           KOYUKUK, AK  United States  65.306667  -157.663611
81666             Kanab, UT  United States    037038N     1123153W
24325       PORT HEIDEN, AK  United States        NaN          NaN
745                NAPA, CA  United States        NaN          NaN
35178          ALVARADO, TX  United States        NaN          NaN
58594         SNOHOMISH, WA  United States  47.863889  -122.068056
44120              EDNA, TX  United States        NaN          NaN
6979        KANSAS CITY, KS  United States        NaN          NaN
61059         Hillsboro, OR  United States  45.540277  -122.949722
66240          New York, NY  United States    404457N     0735958W
40403      EDEN PRAIRIE, MN  United States        NaN          NaN
42901   NORTH LAS VEGAS, NV  United States        NaN          NaN
66417  Porto Segura, Brazil         Brazil    163227S     0039628W
30056          BARNWELL, SC  United States        NaN          NaN
73513   Osaka, Japan, Japan          Japan        NaN          NaN
49523        INDIANTOWN, FL  United States        NaN          NaN
45520     CREAL SPRINGS, IL  United States        NaN          NaN
47215            CONROE, TX  United States        NaN          NaN
61135            HARMON, IL  United States  41.758056     -89.5825
58609           Rome, Italy          Italy        NaN          NaN
57722           ABILENE, TX  United States  32.411111   -99.681667
41957     DAYTONA BEACH, FL  United States        NaN          NaN
71153          McKinney, TX  United States    331230N     0964430W
45098        BAINBRIDGE, GA  United States        NaN          NaN
38169            JUNEAU, AK  United States        NaN          NaN
88482           Ivanhoe, NC  United States    034354N     0781431W
81918            Dayton, OH  United States    393520N     0841329W
24192       KING SALMON, AK  United States        NaN          NaN
45511        YOUNGSTOWN, OH  United States        NaN          NaN
65039          Cheyenne, WY  United States    004190N     0104490W